### PR TITLE
feat: Agent

### PR DESCRIPTION
**Pull request by simster7**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: cmd/argoexec/commands/root.go [hunk 1]
```diff
@@ -77,10 +79,10 @@ func NewRootCommand() *cobra.Command {
 func initExecutor() *executor.WorkflowExecutor {
 	version := argo.GetVersion()
 	executorType := os.Getenv(common.EnvVarContainerRuntimeExecutor)
-	log.WithFields(log.Fields{"version": version.Version, "executorType": executorType}).Info("Starting Workflow Executor")
+	log.WithFields(log.Fields{"version": version, "executorType": executorType}).Info("Starting Workflow Executor")
 	config, err := clientConfig.ClientConfig()
+	config = restclient.AddUserAgent(config, fmt.Sprintf("argo-workflows/%s executor/%s", version.Version, executorType))
 	checkErr(err)
-	config = restclient.AddUserAgent(config, fmt.Sprintf("argo-workflows/%s argo-executor/%s", version.Version, executorType))
```
Comment (alexec) [inline cmd/argoexec/commands/root.go:83]: why did this line change?

File: workflow/common/common.go [hunk 2]
```diff
@@ -105,6 +105,8 @@ const (
 
 	// EnvVarPodName contains the name of the pod (currently unused)
 	EnvVarPodName = "ARGO_POD_NAME"
+	// EnvVarWorkflowName
+	EnvVarWorkflowName = "ARGO_WORKFLOW_NAME"
```
Comment (alexec) [inline workflow/common/common.go:109]: 👍

File: workflow/executor/agent.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+						result.Outputs = outputs
+					}
+					if obj.Status == nil {
+						obj.Status = &wfv1.WorkflowTaskSetStatus{}
+					}
+					if obj.Status.Nodes == nil {
+						obj.Status.Nodes = map[string]wfv1.NodeResult{}
+					}
+					obj.Status.Nodes[task.NodeID] = result
+					log.WithField("obj", obj).Info("will update obj")
```
Comment (alexec) [inline workflow/executor/agent.go:61]: this logging is not clear to me

File: workflow/executor/agent.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+					ts, err := taskSetInterface.UpdateStatus(ctx, obj, metav1.UpdateOptions{})
+					if err != nil {
+						return err
+					}
+
+					log.WithField("taskset", ts).Info("got back task set")
```
Comment (alexec) [inline workflow/executor/agent.go:67]: just logged for diagnostics?

File: workflow/executor/agent.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+func (we *WorkflowExecutor) executeHTTPTemplate(ctx context.Context, tmpl wfv1.Template) (*wfv1.Outputs, error) {
+	httpTemplate := tmpl.HTTP
+	request, err := http.NewRequest(httpTemplate.Method, httpTemplate.URL, bytes.NewBuffer(httpTemplate.Body))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, header := range httpTemplate.Headers {
+		value := header.Value
+		if header.ValueFrom != nil || header.ValueFrom.SecretKeyRef != nil {
```
Comment (alexec) [inline workflow/executor/agent.go:85]: nice

File: workflow/executor/agent.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+	log.WithFields(log.Fields{"url": request.URL, "status": out.Status}).Info("HTTP request made")
+	if out.StatusCode >= 300 {
+		return nil, fmt.Errorf(out.Status)
+	}
+
+	data, err := ioutil.ReadAll(out.Body)
+	if err != nil {
+		return nil, err
+	}
```
Comment (alexec) [inline workflow/executor/agent.go:108]: must close http body

File: workflow/executor/executor.go [hunk 7]
```diff
@@ -90,14 +92,17 @@ type Initializer interface {
...
 	// CopyFile copies a source file in a container to a local path
 	CopyFile(containerName, sourcePath, destPath string, compressionLevel int) error
 
 	// GetOutputStream returns the entirety of the container output as a io.Reader
 	// Used to capture script results as an output parameter, and to archive container logs
 	GetOutputStream(ctx context.Context, containerName string, combinedOutput bool) (io.ReadCloser, error)
 
+	// GetExitCode returns the exit code of the container
+	// Used to capture script exit code as an output parameter
+	GetExitCode(ctx context.Context, containerName string) (string, error)
```
Comment (alexec) [inline workflow/executor/executor.go:104]: we should be getting rid of this, merge issue?

File: workflow/executor/executor.go [hunk 8]
```diff
@@ -710,6 +716,24 @@ func (we *WorkflowExecutor) CaptureScriptResult(ctx context.Context) error {
 	return nil
 }
 
+// CaptureScriptExitCode will add the exit code of a script template as output exit code
+func (we *WorkflowExecutor) CaptureScriptExitCode(ctx context.Context) error {
```
Comment (alexec) [inline workflow/executor/executor.go:720]: again, not sure why this is in this PR?

File: workflow/executor/executor.go [hunk 9]
```diff
@@ -720,28 +744,53 @@ func (we *WorkflowExecutor) AnnotateOutputs(ctx context.Context, logArt *wfv1.Ar
...
+
+	if err := we.annotatePodWithOutputs(ctx, outputs); !apierr.IsForbidden(err) { // me were either successful (nil) or some other error
+		return err
+	}
+
+	return we.updateTaskSetStatusWithOutputs(ctx, outputs)
+}
+
+func (we *WorkflowExecutor) updateTaskSetStatusWithOutputs(ctx context.Context, outputs *wfv1.Outputs) error {
+	log.Info("Patching taskset status with outputs")
```
Comment (alexec) [inline workflow/executor/executor.go:756]: maybe log the outputs?

File: workflow/executor/executor.go [hunk 10]
```diff
@@ -720,28 +744,53 @@ func (we *WorkflowExecutor) AnnotateOutputs(ctx context.Context, logArt *wfv1.Ar
...
+	tasksetInterface := we.workflowInterface.ArgoprojV1alpha1().WorkflowTaskSets(we.Namespace)
+	return waitutil.Backoff(ExecutorRetry, func() (bool, error) {
+		taskset, err := tasksetInterface.Get(ctx, we.workflowName, metav1.GetOptions{})
+		if err != nil && !errorsutil.IsTransientErr(err) {
+			return false, err
+		}
+		if taskset.Status == nil {
```
Comment (alexec) [inline workflow/executor/executor.go:763]: repeated logic here in a hard to test place, instead can we add a `taskSet.SetNodeResult()` func?

File: workflow/executor/executor.go [hunk 11]
```diff
@@ -720,28 +744,53 @@ func (we *WorkflowExecutor) AnnotateOutputs(ctx context.Context, logArt *wfv1.Ar
...
 	if err != nil {
-		return errors.InternalWrapError(err)
+		return err
 	}
-	return we.AddAnnotation(ctx, common.AnnotationKeyOutputs, string(outputBytes))
+	return we.AddAnnotation(ctx, common.AnnotationKeyOutputs, string(data))
 }
 
-// AddError adds an error to the list of encountered errors during execution
+// AddError adds an error to the list of encountered errors durign execution
```
Comment (alexec) [inline workflow/executor/executor.go:788]: spelling

File: workflow/executor/executor.go [hunk 12]
```diff
@@ -720,28 +744,53 @@ func (we *WorkflowExecutor) AnnotateOutputs(ctx context.Context, logArt *wfv1.Ar
...
 func (we *WorkflowExecutor) AddError(err error) {
 	log.Errorf("executor error: %+v", err)
 	we.errors = append(we.errors, err)
 }
 
-// HasError return the first error if exist
-func (we *WorkflowExecutor) HasError() error {
-	if len(we.errors) > 0 {
```
Comment (alexec) [inline workflow/executor/executor.go:739]: merge issue?

File: workflow/executor/executor.go [hunk 13]
```diff
@@ -942,17 +991,14 @@ func watchFileChanges(ctx context.Context, pollInterval time.Duration, filePath
...
+			if err != nil {
 				log.Fatal(err)
-			} else {
-				newModTime := file.ModTime()
-				if modTime != nil && !modTime.Equal(file.ModTime()) {
-					res <- struct{}{}
-				}
-				modTime = &newModTime
 			}
+			newModTime := file.ModTime()
```
Comment (alexec) [inline workflow/executor/executor.go:997]: merge issue?

File: workflow/executor/executor.go [hunk 14]
```diff
@@ -1084,10 +1130,10 @@ func (we *WorkflowExecutor) KillSidecars(ctx context.Context) error {
 			sidecarNames = append(sidecarNames, n)
 		}
 	}
-	log.Infof("Killing sidecars %q", sidecarNames)
 	if len(sidecarNames) == 0 {
 		return nil // exit early as GetTerminationGracePeriodDuration performs `get pod`
 	}
+	log.Infof("Killing sidecars %s", strings.Join(sidecarNames, ","))
```
Comment (alexec) [inline workflow/executor/executor.go:1136]: merge issue?

File: workflow/executor/resource.go [hunk 15]
```diff
@@ -13,24 +15,22 @@ import (
...
 	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	"github.com/argoproj/argo-workflows/v3/util"
-	envutil "github.com/argoproj/argo-workflows/v3/util/env"
 	argoerr "github.com/argoproj/argo-workflows/v3/util/errors"
 	os_specific "github.com/argoproj/argo-workflows/v3/workflow/executor/os-specific"
 )
 
 // ExecResource will run kubectl action against a manifest
-func (we *WorkflowExecutor) ExecResource(action string, manifestPath string, flags []string) (string, string, string, error) {
+func (we *WorkflowExecutor) ExecResource(action string, manifestPath string, flags []string) (string, string, error) {
```
Comment (alexec) [inline workflow/executor/resource.go:30]: merge issue?

File: workflow/executor/resource.go [hunk 16]
```diff
@@ -40,29 +40,22 @@ func (we *WorkflowExecutor) ExecResource(action string, manifestPath string, fla
...
-	resourceName := obj.GetName()
-	resourceKind := obj.GroupVersionKind().Kind
-	if resourceName == "" || resourceKind == "" {
-		return "", "", "", errors.New(errors.CodeBadRequest, "Kind and name are both required but at least one of them is missing from the manifest")
-	}
-	resourceFullName := fmt.Sprintf("%s.%s/%s", resourceKind, resourceGroup, resourceName)
-	selfLink := obj.GetSelfLink()
-	log.Infof("Resource: %s/%s. SelfLink: %s", obj.GetNamespace(), resourceFullName, selfLink)
-	return obj.GetNamespace(), resourceFullName, selfLink, nil
+	resourceName := fmt.Sprintf("%s.%s/%s", obj.GroupVersionKind().Kind, obj.GroupVersionKind().Group, obj.GetName())
```
Comment (alexec) [inline workflow/executor/resource.go:56]: merge issue?

File: workflow/executor/resource.go [hunk 17]
```diff
@@ -1,6 +1,8 @@
 package executor
```
Comment (alexec) [inline workflow/executor/resource.go:1]: merge issue whole file should be reverted?

File: workflow/executor/agent.go [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+			secret, err := util.GetSecrets(ctx, we.ClientSet, we.Namespace, header.ValueFrom.SecretKeyRef.Name, header.ValueFrom.SecretKeyRef.Key)
+			if err != nil {
+				return nil, err
+			}
+			value = string(secret)
+		}
+		request.Header.Add(header.Name, value)
+	}
+
+	out, err := http.DefaultClient.Do(request)
```
Comment (jessesuen) [inline workflow/executor/agent.go:95]: https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

**Pull request closed without merging**